Clean up graph traversal code

edsko opened the pull request:
This pull request is the second of a set of pull requests working towards an implementation of setup dependencies, and depends on #2487 .

There is a lot of duplication between Cabal and cabal-install, and in fact, a lot of dead code in cabal-install. There is still duplication, but no dead code anymore, and at least the situation is clarified now. Details in the commit messages.

/cc @dcoutts @kosmikus

23Skidoo commented:
Build fails on Travis, it looks like you forgot to update `exposed-modules` in `cabal-install.cabal`.

edsko commented:
I am assuming you mean `other-modules`; sorry about that. Added now.

23Skidoo commented:
/cc @ezyang

ezyang commented:
Oof, now I feel a bit bad about myself, because the Cabal/cabal-install package index unification was my fault, and it looks like I did get it wrong (for precisely the reasons described in 1beba1b). It would be good if we had some tests for those dependency cases, they're easy to forget about.

One thing I would like to know (for my own curiosity): does this patchset fix bugs, or is it purely refactoring? It sure looks like it fixes bugs...

edsko commented:
@ezyang So far in this patch set no real behavioural changes, I think -- except from "Avoid forgetting installed IDs" https://github.com/edsko/cabal/commit/daecdef95e52dfef2cb5c0f865bf00cf773dd8bc perhaps, but even there we were okay so far because the roundtrip installed ID -> fake ID -> installed ID is lossless as long as the single instance restriction applies. 

Real changes coming up soon in the next patch set, and that will also include (the start of) a unit test suite for the solver.

23Skidoo commented:
> There is still duplication, but no dead code anymore

Looks like various `FakeMap`-related functions from `D.S.PackageIndex` (`fakeLookupInstalledPackageId`, `brokenPackages'`, `dependencyClosure'`, ...) are now either no longer used or only used inside `cabal-install`. Are they used inside GHC? If not, can we move all `FakeMap`-related machinery to `PlanIndex`?

ezyang commented:
`FakeMap` machinery was specifically for cabal-install, no they are not used in GHC.

23Skidoo commented:
@ezyang 
Thanks, makes sense.

@edsko 
LGTM modulo some comments. Will wait for a comment from @kosmikus or @dcoutts before merging.

@tibbe 
Could you please give @edsko commit rights to the Cabal repo?

tibbe commented:
@23Skidoo done.

23Skidoo commented:
@tibbe Thanks!

edsko commented:
@23Skidoo I've addressed your comments in https://github.com/edsko/cabal/commit/0ff63413d29e3c99083649fd62fde949ab48869c. This leaves:

* Getting rid of the `FakeMap` stuff from `Cabal`

and potentially (less important, I think) 

* Change the names of `PackageInstalled` and `PackageFixedDeps` to make it more obvious that these classes are related
* Use `ConfiguredId` instead of `InstalledPackageId` throughout
* Use `Data.Set` or `ordNub` in the definition of `resolveInstalledIds` 

I will deal try to remove `FakeMap` from `Cabal`, but I'm keen to get the rest of my changes merged in as well, all the above is just preliminary to the real work :) So I suggest I submit those other two PRs first and then once we're all happy with that I will try to remove `FakeMap` from `Cabal`. I suggest we leave the rest until later. 

As regards this comment:

> brokenPackages is just copy-pasted brokenPackages' from D.S.PackageIndex, correct? And analogously for the rest of the functions in this module that now accept FakeMap?

Yes, so far, but that will change in the next PR (or perhaps the one after that) once we start changing `PackageFixedDeps` to actually have more fine-grained dependencies.

edsko commented:
Uh, no idea why Travis is failing, I didn't change nuthin' :)

23Skidoo commented:
Sorry, the Travis breakage was my fault. Should be fixed now.

> So I suggest I submit those other two PRs first and then once we're all happy with that I will try to remove `FakeMap` from `Cabal`. I suggest we leave the rest until later.

Sure, if that's your preference. You may want to create a ticket and assign yourself to it so that you don't forget about this.

edsko commented:
@23Skidoo, @ezyang Ok, removed `FakeMap` completely from `Cabal`, it's now only used in `cabal-install`. So are we okay to merge this PR (and #2487), pending OK from @kosmikus or @dcoutts ?

23Skidoo commented:
@edsko 
> So are we okay to merge this PR (and #2487), pending OK from @kosmikus or @dcoutts ?

Yes, everything looks good to me.

kosmikus commented:
I'm ok with this.

edsko commented:
@23Skidoo Since @kosmikus is okay with this, will we merge?

23Skidoo commented:
OK, let's merge. Potential comments from @dcoutts will be addressed later.

edsko commented:
Thanks!

Note that @dcoutts will be unavailable for the next two weeks but I have discussed this PR, as well as the most of the ones that will follow, with him in detail.
